mobile/examples/phi-3/android/app/src/main/java/ai/onnxruntime/genai/demo/MainActivity.java

```diff
@@ -131,77 +141,57 @@ public void onClick(View v) {
                 new Thread(new Runnable() {
                     @Override
                     public void run() {
-                        TokenizerStream stream = null;
-                        GeneratorParams generatorParams = null;
-                        Generator generator = null;
-                        Sequences encodedPrompt = null;
                         try {
-                            stream = tokenizer.createStream();
-
-                            generatorParams = model.createGeneratorParams();
-                            //examples for optional parameters to format AI response
+                            // Create generator parameters
+                            GeneratorParams generatorParams = genAI.createGeneratorParams();
+                            
+                            // Set optional parameters to format AI response
                             // https://onnxruntime.ai/docs/genai/reference/config.html
-                            generatorParams.setSearchOption("length_penalty", lengthPenalty);
-                            generatorParams.setSearchOption("max_length", maxLength);
-
-                            encodedPrompt = tokenizer.encode(promptQuestion_formatted);
-                            generatorParams.setInput(encodedPrompt);
-
-                            generator = new Generator(model, generatorParams);
-
-                            // try to measure average time taken to generate each token.
+                            generatorParams.setSearchOption("length_penalty", (double)lengthPenalty);
+                            generatorParams.setSearchOption("max_length", (double)maxLength);
                             long startTime = System.currentTimeMillis();
-                            long firstTokenTime = startTime;
-                            long currentTime = startTime;
-                            int numTokens = 0;
-                            while (!generator.isDone()) {
-                                generator.computeLogits();
-                                generator.generateNextToken();
-                 
-                                int token = generator.getLastTokenInSequence(0);
-
-                                if (numTokens == 0) { //first token
-                                    firstTokenTime = System.currentTimeMillis();
+                            AtomicLong firstTokenTime = new AtomicLong(startTime);
+                            AtomicInteger numTokens = new AtomicInteger(0);
+                            
+                            // Token listener for streaming tokens
+                            Consumer<String> tokenListener = token -> {
+                                if (numTokens.get() == 0) { // first token
+                                    firstTokenTime.set(System.currentTimeMillis());
```

Comment:
reading and then writing two atomic variables is not thread-safe, even though the variables themselves are atomic. there's a [compareAndSet](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicLong.html#compareAndSet-long-long-) that might be better suited for this. e.g.:

```suggestion
                                firstTokenTime.compareAndSet(startTime, System.currentTimeMillis());
```

alternatively, if we can get some guarantee from the API that the listener will only be called from a single thread, as it is [now](https://github.com/microsoft/onnxruntime-genai/blob/b4740f1689058567f814c697e92cfc6e0895142b/src/java/src/main/java/ai/onnxruntime/genai/SimpleGenAI.java#L79-L83), we could avoid any thread-safety precautions entirely.